README
Logic-TruthTable, version 1.00, 2 December 2016
-------------------------------------------------------------------------------

A structure to hold a Boolean truth table. Provides utility functions
to create or manipulate the minterms or maxterms of the table columns,
and makes use of the appropriate Algorithm modules for solving equations.

INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	Build
	Build test
	Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Logic::TruthTable

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Logic-TruthTable

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Logic-TruthTable

    CPAN Ratings
        http://cpanratings.perl.org/d/Logic-TruthTable

    Search CPAN
        http://search.cpan.org/dist/Logic-TruthTable/


LICENSE AND COPYRIGHT

Copyright (C) 2016 John M. Gamble

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.



Changes
Revision history for Boolean-TruthTable

1.00    2016 November 5
        First version, released on an unsuspecting world.



Build.PL
use 5.010001;
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
	module_name => 'Logic::TruthTable',

	dist_author => q{John M. Gamble <jgamble@cpan.org>},
	dist_abstract => q(Create a Boolean Truth Table),
	dist_version => '1.00',
	release_status => 'stable',

	provides => {
		'Logic::TruthTable' => {
			file => 'lib/Logic/TruthTable.pm',
		},
		'Logic::TruthTable::Util' => {
			file => 'lib/Logic/TruthTable/Util.pm',
		},
	},

	requires => {
		perl => '5.10.1',
		Moose => '2.16',
		Carp => 0,
		JSON => '2.61',
		'Logic::Minimizer' => '1.00',
		'List::MoreUtils' => '0.401',
		'Module::Runtime' => '0.014',
		'Text::CSV' => 0,
		'namespace::autoclean' => '0.16',
	},
	configure_requires => {
		'Module::Build' => '0.4',
	},
	build_requires => {
		'Test::More' => 0,
		'Test::Deep' => '1.12',
	},

	license => 'perl',
	create_license => 1,
	create_readme => 0,
	create_makefile_pl => 'traditional',
	dynamic_config  => 0,
	add_to_cleanup	 => [ 'Logic-TruthTable-*' ],

	meta_merge => {
		keywords => [qw(boolean boolean-algebra logic truth-table)],

		resources => {
			repository => 'git://github.com/jgamble/Logic-TruthTable',
#			repository => {
#				url => 'git://github.com/jgamble/Logic-TruthTable',
#				web => 'https://github.com/jgamble/Logic-TruthTable',
#				type => 'git',
#			},
		},
	},
);

$builder->create_build_script();


lib/Logic/TruthTable/Util.pm
=head1 NAME

Logic::TruthTable::Util - provide utility functions to Logic::TruthTable

=cut

package Logic::TruthTable::Util;

use strict;
use warnings;
use 5.010001;

use Carp;
use Exporter;
our @ISA = qw(Exporter);

our %EXPORT_TAGS = (
	all => [ qw(
		push_minterm_columns
		push_maxterm_columns
		var_column
		shift_terms
		rotate_terms
		reverse_terms
	) ],
);

our @EXPORT_OK = (
	@{$EXPORT_TAGS{all}},
);

our $VERSION = 1.00;

=head1 DESCRIPTION

This module provides various utilities designed for (but not limited to)
creating or manipulating term lists for Logic::TruthTable.

=cut

=head2 FUNCTIONS

=head3 push_minterm_columns()

=head3 push_maxterm_columns()

    push_minterm_columns($idx, $dir, \@colx, \@coly, \@colz);

or

    push_maxterm_columns($idx, $dir, \@colx, \@coly, \@colz);

Often the outputs to be simulated by boolean expressions are values that
are split across more than one column. For example, say that you want
to model a function to direct a pointer that uses the eight
L<cardinal and ordinal |https://en.wikipedia.org/wiki/Points_of_the_compass#Compass_point_names>
compass directions, from North (value 0) to NorthWest (value 7).

Numbering these directions takes three bits, which means you'd need three
columns to represent them.

To make it easier to create these columns, C<push_minterm_columns()>
(or, if you prefer, C<push_maxterm_columns()>) will take a value
from your function and, for each set bit (or if using maxterms,
unset bit), will push the minterm (or maxterm) onto each array
corresponding to its column.

For example, if the value of row 20 is 13 (in binary C<0b1101>),
then a call to C<push_minterm_umns(20, 13, \@w, \@x, \@y, \@z);>
will push 20 onto array variables C<@w>, C<@x>, and C<@z>, while a call to
C<push_maxterm_umns(20, 13, \@w, \@x, \@y, \@z);> will push a 20 onto
array variable C<@y> only.

Bit values past the available columns will simply be dropped, while
excess columns will either either never have terms pushed on them
(C<push_minterm_columns()>) or always have terms pushed on them
(C<push_maxterm_columns()>).

For example:

    #
    # Each column gets its own term list.
    # The don't-care terms will be common across
    # all columns.
    #
    my(@colx, @coly, @colz, @dontcares);

    #
    # For each cell, return a direction.
    #
    for my $idx (0..63)
    {
        my $dir = sp_move_to($idx);

        #
        # In this example, a cell that cannot be exited cannot
        # be entered either, so mark it as a don't-care.
        #
        if ($dir < 0 or $dir > 7)
        {
            push @dontcares, $idx;
        }
        else
        {
            #
            # For any set bit in $dir, push $idx onto the corresponding
            # column list.
            #
            push_minterm_columns($idx, $dir, \@colx, \@coly, \@colz);
        }
    }

You will then have the minterms available for each column of your
truth table.

    my $dir_table = Logic::TruthTable->new(
                title => "Sandusky Path",
                width => 6,
                vars => [qw(A B C D E F)],
                functions => [qw(X Y Z)],
                columns => [
                    {
                        minterms => \@colx,
                        dontcares => \@dontcares,
                    },
                    {
                        minterms => \@coly,
                        dontcares => \@dontcares,
                    },
                    {
                        minterms => \@colz,
                        dontcares => \@dontcares,
                    } ],
        );

In some cases you may not even know how many columns of output (or input)
you will need for your problem. But since the vars, functions, and columns
attributes are all array references, these can be created programatically:

    #
    # Create a truth table for an extended Rock Paper Scissors game.
    #
    my $tt = extended_rps(7);

    #
    # We may not have the faster algorthm installed, and the
    # truth table may be very large, so save the table as a
    # file for Logic Friday.
    #
    $tt->export_csv(dc => 'X', write_handle => \*STDOUT);
    exit(0);

    #
    # 'Winner' function for an extended Rock Paper Scissors game.
    #
    sub rpswinner
    {
        my($player_a, $player_b, $length) = @_;

        return 0 if ($player_a == $player_b);    # Tie, of course.

        my $val = $player_a - $player_b + $length;

        return $player_a if ($val % 3 == 1);
        return $player_b;
    }

    #
    # Create an extended Rock Paper Scissors 'winners' table.
    # Given two throws, return which of the two throws wins.
    # For example, if the two throws in a game of
    # Rock Paper Scissors Spock Lizard are Paper and Lizard, the
    # output columns would indicate Lizard (eats Paper).
    #
    # Throws are numbers from 1 (Rock) onward, with Tie being 0.
    #
    # Returns the Logic::TruthTable object.
    #
    sub extended_rps
    {
        my($n) = @_;
        my @dontcares;

        #
        # How many binary columns do we need represent a throw?
        #
        my $w = length(sprintf("%b", $n));

        my $two_exp_w = 2 ** $w;
        my $length = $two_exp_w - 1;

        #
        # Make an empty array of output terms.
        #
        my @aterms;
        push @aterms, [] for (0 .. $w-1);

        #
        # Now, all the possible combinations of player A vs. player B.
        #
        for my $j (0 .. $length)
        {
            for my $k (0 .. $length)
            {
                my $row = $j * $two_exp_w + $k;
                if ($k == 0 or $j == 0)
                {
                    push @dontcares, $row;
                }
                else
                {
                    my $val = rpswinner($j, $k, $length);
                    push_minterm_columns($row, $val, @aterms);
                }
            }
        }
        #
        # Let's create the table's columns.
        #
        my @columns;

        for my $idx (0 .. $w-1)
        {
            push @columns, {dontcares => [@dontcares],
                            minterms => [@{$aterms[$idx]}], };
        }

        #
        # And the variable and function names.
        # (We are assuming fewer than 1024 choices here.)
        #
        my @vars_a = reverse(('a0' .. 'a9')[0 .. $w-1]);
        my @vars_b = reverse(('b0' .. 'b9')[0 .. $w-1]);
        my @fns = reverse(('w0' .. 'w9')[0 .. $w-1]);

        my $tt = Logic::TruthTable->new(
            title => "$n-throw Rock Paper Scissors (extended)",
            width =>  2 * $w,
            vars => [@vars_a, @vars_b],
            functions => [@fns],
            columns => [@columns],
        );

        return $tt;
    }


=cut

sub push_minterm_columns
{
	my($idx, $val, @colrefs) = @_;

	my $ncols = $#colrefs;
	my $bit = 1 << $ncols;

	#
	# Slice the bits across the columns.
	#
	for my $j (0..$ncols)
	{
		push @{ $colrefs[$j] }, $idx if ($val & $bit);
		$bit >>= 1;
	}
}

sub push_maxterm_columns
{
	my($idx, $val, @colrefs) = @_;

	my $ncols = $#colrefs;
	my $bit = 1 << $ncols;

	#
	# Slice the bits across the columns.
	#
	for my $j (0..$ncols)
	{
		push @{ $colrefs[$j] }, $idx unless ($val & $bit);
		$bit >>= 1;
	}
}

=head3 var_column()

Return the list of terms that correspond to the set bits of a
variable's column.

    my @terms = var_column($width, $col);

For example, in a three-variable table

       x  y  z  | f
     -----------|--
  0  | 0  0  0  |
  1  | 0  0  1  |
  2  | 0  1  0  |
  3  | 0  1  1  |
  4  | 1  0  0  |
  5  | 1  0  1  |
  6  | 1  1  0  |
  7  | 1  1  1  |

column 2 (the x column) has terms (4, 5, 6, 7) set, while column 0
(the z column) has terms (1, 3, 5, 7) set.

=cut

sub var_column
{
	my($width, $col) = @_;

	croak "Column $col doesn't exist in a $width-column table." if ($col >= $width);

	#
	# A 'block' is a sequence of ones in the column.
	# 'blocklen' is the number of ones, blocks is the number of
	# those sequences.
	#
	# So in a set of four variable columns, column one's ones come
	# in eight sets of two ones.
	#
	my $blocklen = 1 << $col;
	my $blocks = 1 << ($width - $col - 1);
	my @terms;

	for my $n (0 .. $blocks - 1)
	{
		push @terms, map { (2 * $n + 1) * $blocklen + $_} (0 .. $blocklen - 1);
	}

	return @terms;
}

=head3 reverse_terms()

Reverses the list of terms by index. For example, within a four-bit range:

    $width = 4;                  # values range 0 .. 15.
    @terms = (1, 3, 6, 8, 13, 14);
    @terms = reverse_terms($width, \@terms);

The values in C<@terms> will become (14, 12, 9, 7, 2, 1).

=cut

sub reverse_terms
{
	my($width, $tref) = @_;
	my $last = (1 << $width) - 1;

	return map {$last - $_} @{ $tref };
}

=head3 rotate_terms()

Rotates the list of terms by index. For example, within a four-bit range:

    $width = 4;                  # values range 0 .. 15.
    $shift = 5;                  # term 0 becomes term 5, term 1 becomes term 6
    @terms = (1, 3, 7, 9, 13, 15);
    @terms = rotate_terms($width, \@terms, $shift);

The values in C<@terms> will become will be (6, 8, 12, 14, 2, 4),
with the last two list items rotated around to the beginning, having
been rotated past what C<$width> allows. A negative-valued shift rotates
the terms backward.

=cut

sub rotate_terms
{
	my($width, $tref, $shift) = @_;
	my $length = 1 << $width;

	$shift %= $length;

	$shift += $length while ($shift < 0);

	return @{ $tref } if ($shift == 0);

	return map {($shift + $_) % $length} @{ $tref };
}

=head3 shift_terms()

Shifts the list of terms by index. For example, within a four-bit range:

    $width = 4;                  # values range 0 .. 15.
    $shift = 5;                  # term 0 becomes term 5, term 1 becomes term 6
    @terms = (1, 3, 7, 9, 13, 15);
    @terms = shift_terms($width, \@terms, $shift);

The values in C<@terms> will become (6, 8, 12, 14), with the last
two list items dropped, having been shifted past what C<$width> allows.
A negative-valued shift shifts the terms downward.

=cut

sub shift_terms
{
	my($width, $tref, $shift) = @_;
	my $length = 1 << $width;

	if ($shift >= $length or $shift <= -$length)
	{
		return ();
	}
	elsif ($shift > 0)
	{
		return map {$shift + $_} grep {$_ < ($length - $shift)} @{ $tref };
	}
	else
	{
		return map {$shift + $_} grep {$_ >= -$shift} @{ $tref };
	}
}

=head1 SEE ALSO

L<Logic::TruthTable::Base81>

=head1 AUTHOR

John M. Gamble C<< <jgamble@cpan.org> >>

=cut

1;

__END__



lib/Logic/TruthTable.pm
package Logic::TruthTable;

use 5.010001;
use strict;
use warnings;

use Moose;
use Moose::Util::TypeConstraints;
use namespace::autoclean;

use Carp;
use List::MoreUtils qw(pairwise);
use Module::Runtime qw(is_module_name use_module);
use Text::CSV;

use Logic::Minimizer;

#
# TBD: Parallelize the column-solving. Some
# recommended modules below, choose later.
#
#use Parallel::ForkManager;
#use MCE;
#use Smart::Comments ('###'); 

#
# Base class of the minimizer (used by
# Algorithm::QuineMcCluskey).
#
class_type 'ColumnMinimizer',
	{class => 'Logic::Minimizer'};

#
# Define the array of minimizer/hashref types
# used to define our table's function columns.
#
subtype 'ArrayRefOfColumnMinimizers',
	as 'ArrayRef[ColumnMinimizer]';

subtype 'ArrayRefOfHashRef',
	as 'ArrayRef[HashRef]';

#
# The width attribute is fed into the
# minimizer object; it cannot be overridden
# by the minimizer's attributes.
#
has 'width' => (
	isa => 'Int', is => 'ro', required => 1
);

#
# The don't-care character and the vars attributes on the
# other hand, are merely defaults and *can* be overridden
# by the object.
#
has 'dc' => (
	isa => 'Str', is => 'rw',
	default => '-'
);

has 'vars' => (
	isa => 'ArrayRef[Str]', is => 'rw', required => 0,
	default => sub{['A' .. 'Z'];},
);

#
# Used to determine which minimizer object type will be
# created by default for the columns. As of this release,
# only Algorithm::QuineMcCluskey is available.
#
has 'algorithm' => (
	isa => 'Str', is => 'ro', required => 0,
	default => 'QuineMcCluskey',
);

#
# The function names for each column.
#
has 'functions' => (
	isa => 'ArrayRef[Str]', is => 'rw', required => 0,
	default => sub{['F0' .. 'F9', 'F10' .. 'F31'];},
);

#
# The column objects. Either the array ref of hash refs (i.e., the plain
# text), or the algorithm object.
#
has 'columns' => (
	isa => 'ArrayRefOfHashRef|ArrayRefOfColumnMinimizers',
	is => 'ro', required => 1,
	reader => '_get_columns',
	writer => '_set_columns',
	predicate => 'has_columns'
);

#
# The title of the truth table.
#
has 'title' => (
	isa => 'Str', is => 'rw', required => 0,
	predicate => 'has_title'
);

#
# Number of columns (functions). Stored so that we don't have to
# go nuts with array sizing an array reference in an object.
#
has '_fn_width' => (
	isa => 'Int', is => 'rw', required => 0
);

#
# Hash look-up by name instead of by index for column (function)
# or var column.
#
has ['_fn_lookup', '_var_lookup'] => (
	isa => 'HashRef', is => 'rw', required => 0,
);

=head1 NAME

Logic::TruthTable - Create and solve sets of boolean equations.

=head1 VERSION

Version 1.00

=cut

our $VERSION = '1.00';


=head1 SYNOPSIS

Create a truth table.

 
    #
    # Rock-Paper-Scissors winners table.
    #
    # Returns (in two bits) the winner of Rock (01) vs. Paper (10)
    # or vs. Scissors (11). A tie is 00.
    #
    # 
    #        a1 a0 b1 b0 ||  w1 w0
    #       -----------------------
    #  0     0  0  0  0  ||  -  -
    #  1     0  0  0  1  ||  -  -
    #  2     0  0  1  0  ||  -  -
    #  3     0  0  1  1  ||  -  -
    #  4     0  1  0  0  ||  -  -
    #  5     0  1  0  1  ||  0  0    (tie)
    #  6     0  1  1  0  ||  1  0    (paper)
    #  7     0  1  1  1  ||  0  1    (rock)
    #  8     1  0  0  0  ||  -  -
    #  9     1  0  0  1  ||  1  0    (paper)
    # 10     1  0  1  0  ||  0  0    (tie)
    # 11     1  0  1  1  ||  1  1    (scissors)
    # 12     1  1  0  0  ||  -  -
    # 13     1  1  0  1  ||  0  1    (rock)
    # 14     1  1  1  0  ||  1  1    (scissors)
    # 15     1  1  1  1  ||  0  0    (tie)
    #

    use Logic::TruthTable;

    my $ttbl = Logic::TruthTable->new(
        width => 4,
        title => 'Rock Paper Scissors Winner Results',
        vars => ['a1', 'a0', 'b1', 'b0'],
        functions => ['w1', 'w0'],
        columns => [
            {
                minterms => [6, 9, 11, 14],
                dontcares => [0 .. 4, 8, 12],
            },
            {
                minterms => [7, 11, 13, 14],
                dontcares => [0 .. 4, 8, 12],
            },
        ],
    );

=head1 Description

This module minimizes tables of 
L<Boolean expressions|https://en.wikipedia.org/wiki/Boolean_algebra> using the
algorithms available on CPAN.

=head2 Object Methods

=head3 new()

Create the truth table object. The attributes are:

=over 4

=item 'width'

The number of variables (input columns) in the Boolean expressions.

This is a required attribute.

=item 'title'

A title for the problem you are solving.

=item 'dc'

I<Default value: '-'>

Change the representation of the don't-care character. The don't-care
character is used both in the columnstring, and internally as a place
holder for eliminated variables in the equation. Some of those internals
may be examined via other methods.

This becomes the I<default> value of the function columns; it may be
individually overridden in each C<columns> attribute.

=item 'vars'

I<Default value: ['A' .. 'Z']>

The variable names used to form the equation. The names will be taken from
the leftmost first.

This becomes the I<default> value of the function columns; it may be
individually overridden in each C<columns> attribute.

=item 'functions'

I<Default value: ['F0' .. 'F9', 'F10' .. 'F31']>

The function names used to form the equation.

=item 'algorithm'

The default algorithm that will be used to minimize the Boolean equations.
The name will come from the package name, e.g., having an attribute
C<algorithm => 'QuineMcCluskey'> means that the column will be minimized
using the package Algorithm::QuineMcCluskey.

The algorithm module must be installed, and must be of the form
C<Algorithm::Name>. The module must also have Logic::Minimizer as its
parent class. This ensures that it will have the methods needed by
Logic::TruthTable to create and solve the Boolean expressions.

This becomes the I<default> value of the function columns; it may be
individually overridden in each C<columns> attribute.

=item 'columns'

An array of hash references that create the algorithm minimizer objects.

Each column becomes an object itself, and must have the attributes necessary
to create the object.

Alternatively, it is possible to pre-create the algorithm minimizer objects,
and use them directly in the C<columns> array:

    $q3 = Algorithm::QuineMcCluskey->new(
        title	=> "First column of a four bit binary to 2-4-2-1 converter",
        width => 4,
        minterms => [ 5 .. 9 ],
        dontcares => [ 10 .. 15 ],
        vars => ['w' .. 'z'],
    );
    $q2 = Algorithm::QuineMcCluskey->new(
        title	=> "Second column of a four bit binary to 2-4-2-1 converter",
        width => 4,
        minterms => [ 4, 6 .. 9 ],
        dontcares => [ 10 .. 15 ],
        vars => ['w' .. 'z'],
    );
    $q1 = Algorithm::QuineMcCluskey->new(
        title	=> "Third column of a four bit binary to 2-4-2-1 converter",
        width => 4,
        minterms => [ 2, 3, 5, 8, 9 ],
        dontcares => [ 10 .. 15 ],
        vars => ['w' .. 'z'],
    )
    $q0 = Algorithm::QuineMcCluskey->new(
        title	=> "First column of a four bit binary to 2-4-2-1 converter",
        width => 4,
        minterms => [ 1, 3, 5, 7, 9 ],
        dontcares => [ 10 .. 15 ],
        vars => ['w' .. 'z'],
    )

    #
    # Create the truth table using the above Algorithm::QuineMcCluskey objects.
    #
    $tt_2421 = Logic::TruthTable->new(
        width => 4,
        title	=> "A four bit binary to 2-4-2-1 converter",
        vars => ['w' .. 'z'],
        functions => [qw(a3 a2 a1 a0)],
        columns => [$q3, $q2, $q1, $q0],
    );

=back

=cut

sub BUILD
{
	my $self = shift;
	my $w = $self->width;
	my @cols = @{$self->_get_columns};
	my @fn_names = @{$self->functions};
	my @vars = @{$self->vars};
	my $dc = $self->dc;

	#
	# Make sure the number of function names and variables
	# get set correctly.
	#
	croak "Not enough function names for your columns" if ($#fn_names < $#cols);

	$#fn_names = $#cols;
	$self->functions(\@fn_names);
	$self->_fn_width($#cols);

	$#vars = $w - 1;
	$self->vars(\@vars);
	$self->title("$w-variable truth table in $#cols columns") unless ($self->has_title);

	#
	# Set up the look-up-by-name hashes.
	#
	$self->_fn_lookup({ map{ $fn_names[$_], $_} (0 .. $#fn_names) });
	$self->_var_lookup({ map{ $vars[$_], $_} (0 .. $#vars) });

	#
	# Set up the individual columns, using defaults
	# from the truth table object, if present.
	#
	for my $idx (0 .. $#cols)
	{
		my %tcol = %{ $cols[$idx] };
		$tcol{width} //= $w;
		$tcol{dc} //= $dc;
		$tcol{algorithm} //= $self->algorithm;
		$tcol{vars} //= [@vars];
		$tcol{title} //= $fn_names[$idx];

		croak "Column $idx: width => " . $tcol{width} .
			" doesn't match table's width $w" if ($tcol{width} != $w);

		${$self->_get_columns}[$idx] = new_minimizer_obj(\%tcol);
	}

	return $self;
}

#
# new_minimizer_obj(%algorithm_options)
#
# Creates a column's object (like an Algorithm::QuineMcCluskey object,
# for example) from the options provided.
#
sub new_minimizer_obj
{
	my($href) = @_;
	my %args = %{$href};
	my $al;

	#
	# Find out which object we're creating.
	#
	($al = $args{algorithm}) =~ s/-/::/;
	$al = "Algorithm::" . $al;

	croak "Invalid module name '$al'" unless (is_module_name($al));

	my $obj = use_module($al)->new(%args);
	croak "Couldn't create '$al' object" unless defined $obj;

	return $obj;
}

=head3 get_fncolumn()

Return a column object by name or index.

The columns of a C<Logic::TruthTable> object are themselves
objects, of types C<Algorithm::Name>, where I<Name> is the
algorithm, and which may be set using the C<algorithm> parameter
in C<new()>. (As of this writing, the only algorithm availble
in the CPAN ecosystem is C<Algorithm::QuineMcCluseky>.)

Each column is named via the C<functions> attribute in C<new()>, and
a column can be retrieved using its name.

    my $ttable = Logic::TruthTable->new(
        title => "An Example",
	width => 5,
	functions => ['F1', 'F0'],
        algorithm => 'QuineMcCluskey',  # Our only choice currently.
        columns => [
            {
                minterms => [6, 9, 23, 27],
                dontcares => [0, 2, 4, 16, 24],
            },
            {
                minterms => [7, 11, 19, 23, 29, 30],
                dontcares => [0, 2, 4, 16, 24],
            },
        ],
    );

    my $col_f0 = $ttable->get_fncolumn('F0');

C<$col_f0> will be an Algorithm::QuineMcCluskey object with minterms
(7, 11, 19, 23, 29, 30). As columns are numbered in array order, the
same column could have been retrieved with:

    my $col_f0 = $ttable->get_fncolumn(1);


=cut

sub get_fncolumn
{
	my $self = shift;
	my($fn_name) = @_;
	my $idx;

	#
	#### Let's look at the key: $fn_name
	#### Let's look at the hash: %{$self->_fn_lookup()}
	#### Let's look an an element: $self->_fn_lookup()->{$fn_name}
	#

	$idx = ($self->_fn_lookup()->{$fn_name}) // $fn_name;

	return undef unless ($idx =~ m/^\d+$/ and $idx <= $self->_fn_width);
	return $self->_get_columns()->[$idx];
}

=head3 solve()

=cut

sub solve
{
	my $self = shift;
	my @solns;

	for my $col (@{$self->_get_columns})
	{
		#
		##### Solving: $col
		#
		my $eqn = $col->solve();
		push @solns, $eqn;
	}

	#
	#### solve returns: @solns
	#
	return @solns;
	#return map {$_->solve()} @{$self->_get_columns};
}

=head3 fnsolve()

=cut

sub fnsolve
{
	my $self = shift;
	my(@f) = @{ $self->functions() };
	my(@eqns) = $self->solve();

	return pairwise {qq($a = $b)} @f, @eqns;
}

=head3 export_csv()

Write the truth table out to a CSV file, suitable for reading by
other programs, such as a spreadsheet application, or by
L<Logic Friday|http://sontrak.com/>, a tool for working with logic
functions.

In the example below, a file is being written out for reading
by Logic Friday. Note that Logic Friday insists on its own
don't-care character:

    if (open my $fh, ">:encoding(utf8)", "ttmwc.csv")
    {
        #
        # Override the don't-care character, as Logic Friday
        # insists on it being an 'X'.
        #
        $truthtable->export_csv(write_handle => $fh, dc => 'X');

        close $fh;
    }

The options are:

=over 2

=item write_handle

The opened file handle for writing.

=item dc

The don't-care symbol to use in the file.

=back

The method returns undef if an error is encountered. On
success it returns itself.

Note that this method cannot read Logic Friday's minimized export
format, only the full, not-minimized files.

=cut

sub export_csv()
{
	my $self = shift;
	my(%opts) = @_;

	my $handle = $opts{write_handle};

	### handle: $handle

	unless (defined $handle)
	{
		carp "export_csv(): no file opened for export.";
		return undef;
	}

	my $w = $self->width;
	my $dc = $opts{dc} // $self->dc;
	my $fmt = "%0${w}b";
	my $lastrow = (1 << $w) - 1;
	my @columns;

	#
	# Set up the array of column strings.
	#
	# Set up a potential regex in case the don't-care
	# character is different than the columns' dc.
	#
	### dc: $dc
	#
	for my $c_idx (0 .. $self->_fn_width)
	{
		my $obj = ${$self->_get_columns}[$c_idx];
		my @c = @{$obj->to_columnlist};

		if ($dc ne $obj->dc)
		{
			my $obj_dc = $obj->dc;
			### obj_dc: $obj_dc
			### to dc: $dc
			$_ =~ s/\Q$obj_dc\E/$dc/ for (@c);
		}

		push @columns, [@c];
	}

	#
	# Open the CSV file, print out the header, then each row.
	#
	my $csv = Text::CSV->new( {binary => 1, eol => "\012"} );

	unless ($csv)
	{
		carp "Cannot use Text::CSV: " . Text::CSV->error_diag();
		return undef;
	}

	$csv->print($handle, [@{$self->vars}, '', @{$self->functions}]);

	for my $r_idx (0 .. $lastrow)
	{
		my @row = (split(//, sprintf($fmt, $r_idx)), '');

		push @row, shift @{ $columns[$_] } for (0 .. $self->_fn_width);

		$csv->print($handle, [@row]);
	}

	return $self;
}

=head3 import_csv()

Read a previously written CSV file and create a Logic::TruthTable
object from it.

    #
    # Read in a CSV file.
    #
    if (open my $lf, "<:encoding(utf8)", "excess_3.csv")
    {
        $truthtable = Logic::TruthTable->import_csv(
            read_handle => $lf,
        );
        close $lf;
    }


The don't-care character doesn't have to be set in this case, as import_csv()
will assume anything not a 0 or 1 is the don't-care character. Of course you
may set it anyway if you want to change it from the default don't-care character
in the truth table object.

You can set whether the truth table object is created using the
minterms or the maxterms of the CSV file by using the C<termtype>
attribute:

    $truthtable = Logic::TruthTable->import_csv(
        read_handle => $lf,
        termtype => 'maxterms',        # or 'minterms'.
    );

By default the object is created with minterms.

In addition to the termtype, you may also set the title, don't-care character,
and algorithm attributes. Width, variable names, and function names cannot be
set as these are read from the file.

    $truthtable = Logic::TruthTable->import_csv(
        read_handle => $lf,
        title => "Excess-3 multiplier",
        dc => '.',
        algorithm => 'QuineMcCluskey'
    );

The method returns undef if an error is encountered.

Note that this method cannot read Logic Friday's minimized export
format, only the full, not-minimized files.

=cut

sub import_csv
{
	my $self = shift;
	my(%opts) = @_;

	my $handle = $opts{read_handle};
	my $termtype = $opts{termtype} // 'minterms';

	my @vars;
	my @functions;
	my $width = 0;

	unless (defined $handle)
	{
		carp "import_csv(): no file opened.";
		return undef;
	}
	unless ($termtype =~ /minterms|maxterms/)
	{
		carp "Incorrect value for termtype ('minterms' or 'maxterms')";
		return undef;
	}

	my $csv = Text::CSV->new( {binary => 1} );

	unless ($csv)
	{
		carp "Cannot use Text::CSV: " . Text::CSV->error_diag();
		return undef;
	}

	#
	# Parse the first line of the file, which is the header,
	# and which will have the variable and function names, which
	# in turn will let us deduce the width.
	#
	my $header = $csv->getline($handle);

	#
	### The header is: $header
	#
	for (@$header)
	{
		#
		### Examining: $_
		#
		if ($_ eq '')
		{
			if ($width != 0)
			{
				carp "File is not in the correct format";
				return undef;
			}

			$width = scalar @vars;
		}
		elsif ($width == 0)
		{
			push @vars, $_;
		}
		else
		{
			push @functions, $_;
		}
	}

	#
	# Now that we've got our width, var names, and
	# function names, collect the terms.
	#
	### width: $width
	### termtype: $termtype
	### functions: @functions
	### vars: @vars
	#
	my($termrefs, $dcrefs);

	my $idx = 0;
	while (my $row = $csv->getline($handle))
	{
		for my $c (0 .. $#functions)
		{
			my $field = 1 + $c + $width;

			if ($row->[$field] !~ /[01]/)
			{
				push @{ $dcrefs->[$c] }, $idx;
			}
			elsif (($termtype eq 'minterms' and $row->[$field] eq '1') or
				($termtype eq 'maxterms' and $row->[$field] eq '0'))
			{
				push @{ $termrefs->[$c] }, $idx;
			}
		}
		$idx++;
	}

	#
	# We've collected our variable names, function names, and terms.
	# Let's make an object.
	#
	### dcrefs: $dcrefs
	### termrefs: $termrefs
	#
	my $title = $opts{title} // "$width-input table created from import file";
	my $algorithm = $opts{algorithm} // 'QuineMcCluskey';
	my $dc = $opts{dc} // '-';
	my @columns;

	for my $c (0 .. $#functions)
	{
		push @columns, {
			dontcares => $dcrefs->[$c],
			$termtype, $termrefs->[$c]
		};
	}

	return Logic::TruthTable->new(
		width => $width,
		title => $title,
		dc => $dc,
		vars => [@vars],
		functions => [@functions],
		columns => [@columns],
		algorithm => $algorithm,
	);
}


=head1 AUTHOR

John M. Gamble, C<< <jgamble at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-boolean-truthtable at rt.cpan.org>,
or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Logic-TruthTable>. I will
be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Logic::TruthTable

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Logic-TruthTable>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Logic-TruthTable>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Logic-TruthTable>

=item * Search CPAN

L<http://search.cpan.org/dist/Logic-TruthTable/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2015 John M. Gamble.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut

1; # End of Logic::TruthTable
